File: app/src/main/java/com/hqq/core/app/ui/customize/weight/RatingBarView.java
package com.hqq.core.app.ui.customize.weight;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.BitmapFactory;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.hqq.core.app.R;
import com.hqq.core.utils.ResourcesUtils;
import com.hqq.core.utils.log.LogUtils;

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.core.app.ui.customize.weight
 * @FileName :   RatingBarView
 * @Date : 2019/6/24 0024  下午 4:34
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public class RatingBarView extends LinearLayout {
    public RatingBarView(Context context) {
        this(context, null);
    }

    public RatingBarView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RatingBarView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.RatingBar);
        mIsSlide = array.getBoolean(R.styleable.RatingBar_is_slide, false);
        mMarginRight = (int) array.getDimension(R.styleable.RatingBar_padding_child, 10);

        array.recycle();
        initView();

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        BitmapFactory.Options options = new BitmapFactory.Options();
        BitmapFactory.decodeResource(getResources(), mDefPic, options);
        //获取图片的宽高
        mHeight = options.outHeight < mHeight ? options.outHeight : mHeight;
        mWidth = options.outWidth < mWidth ? options.outWidth : mWidth;
        if (getMeasuredHeight() != 0 && getMeasuredHeight() < mHeight) {
            mWidth = (int) ((float) getMeasuredHeight() / mHeight * mWidth);
            mHeight = getMeasuredHeight();
            initView();
        }
    }

    int mDefSelectPic = R.mipmap.ic_star_red;
    int mDefPic = R.mipmap.ic_star_gray;
    int mCurrent = 0;
    private int mWidth = 50, mHeight = 50;
    int mMarginRight = 10;
    boolean mIsSlide = false;

    private void initView() {

        setBackgroundColor(ResourcesUtils.getColor(getContext(), R.color.white));
        removeAllViews();
        LayoutParams params = new LayoutParams(mWidth, mHeight);
        for (int i = 0; i < 5; i++) {
            ImageView imageView = new ImageView(getContext());
            if (i <= mCurrent) {
                imageView.setImageResource(mDefSelectPic);
            } else {
                imageView.setImageResource(mDefPic);
            }
            if (i < 5) {
                params.setMargins(0, 0, mMarginRight, 0);
            }
            addView(imageView, params);
        }
    }


    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        final float index, length;
        if (getOrientation() == HORIZONTAL) {
            index = ev.getX();
            length = getWidth();
        } else {
            index = ev.getY();
            length = getHeight();
        }
        if (mIsSlide) {

            mCurrent = (int) (5 * index / length);
            LogUtils.e(" index " + mCurrent +"------"+ index);
            redraw();
        }
        return ev.getAction() != MotionEvent.ACTION_UP || super.dispatchTouchEvent(ev);
    }

    private void redraw() {
        for (int i = 0; i < 5; i++) {
            View view = getChildAt(i);
            if (i <= mCurrent) {
                ((ImageView) view).setImageResource(mDefSelectPic);
            } else {
                ((ImageView) view).setImageResource(mDefPic);
            }
        }
    }

    public void setDefSelectPic(int defSelectPic) {
        mDefSelectPic = defSelectPic;
        initView();
    }

    public void setSelect(int select) {
        mCurrent = select;
    }

    public int getCurrent() {
        return mCurrent;
    }

    public void setSlide(boolean slide) {
        mIsSlide = slide;
    }

    public void setWidth(int width) {
        mWidth = width;
    }

    public void setHeight(int height) {
        mHeight = height;
    }

    public void setMarginRight(int marginRight) {
        mMarginRight = marginRight;
    }
}

File: app/src/main/java/com/hqq/example/ui/customize/widget/RatingBarView.kt
package com.hqq.example.ui.customize.widget

import android.content.Context
import android.graphics.BitmapFactory
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.ImageView
import android.widget.LinearLayout
import com.hqq.core.utils.ResourcesUtils.getColor
import com.hqq.core.utils.log.LogUtils.e
import com.hqq.example.R

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.core.app.ui.customize.weight
 * @FileName :   RatingBarView
 * @Date : 2019/6/24 0024  下午 4:34
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
class RatingBarView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : LinearLayout(context, attrs, defStyleAttr) {
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val options = BitmapFactory.Options()
        BitmapFactory.decodeResource(resources, mDefPic, options)
        //获取图片的宽高
        mHeight = if (options.outHeight < mHeight) options.outHeight else mHeight
        mWidth = if (options.outWidth < mWidth) options.outWidth else mWidth
        if (measuredHeight != 0 && measuredHeight < mHeight) {
            mWidth = (measuredHeight.toFloat() / mHeight * mWidth).toInt()
            mHeight = measuredHeight
            initView()
        }
    }

    var mDefSelectPic = R.mipmap.ic_star_red
    var mDefPic = R.mipmap.ic_star_gray
    var current = 0
    private var mWidth = 50
    private var mHeight = 50
    var mMarginRight = 10
    var mIsSlide = false
    private fun initView() {
        setBackgroundColor(getColor(context, R.color.white))
        removeAllViews()
        val params = LayoutParams(mWidth, mHeight)
        for (i in 0..4) {
            val imageView = ImageView(context)
            if (i <= current) {
                imageView.setImageResource(mDefSelectPic)
            } else {
                imageView.setImageResource(mDefPic)
            }
            if (i < 5) {
                params.setMargins(0, 0, mMarginRight, 0)
            }
            addView(imageView, params)
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        val index: Float
        val length: Float
        if (orientation == HORIZONTAL) {
            index = ev.x
            length = width.toFloat()
        } else {
            index = ev.y
            length = height.toFloat()
        }
        if (mIsSlide) {
            current = (5 * index / length).toInt()
            e(" index " + current + "------" + index)
            redraw()
        }
        return ev.action != MotionEvent.ACTION_UP || super.dispatchTouchEvent(ev)
    }

    private fun redraw() {
        for (i in 0..4) {
            val view = getChildAt(i)
            if (i <= current) {
                (view as ImageView).setImageResource(mDefSelectPic)
            } else {
                (view as ImageView).setImageResource(mDefPic)
            }
        }
    }

    fun setDefSelectPic(defSelectPic: Int) {
        mDefSelectPic = defSelectPic
        initView()
    }

    fun setSelect(select: Int) {
        current = select
    }

    fun setSlide(slide: Boolean) {
        mIsSlide = slide
    }

    fun setWidth(width: Int) {
        mWidth = width
    }

    fun setHeight(height: Int) {
        mHeight = height
    }

    fun setMarginRight(marginRight: Int) {
        mMarginRight = marginRight
    }

    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.RatingBar)
        mIsSlide = array.getBoolean(R.styleable.RatingBar_is_slide, false)
        mMarginRight = array.getDimension(R.styleable.RatingBar_padding_child, 10f).toInt()
        array.recycle()
        initView()
    }
}